File: backend/runner/pkg/xqldir/dir.go
package xqldir

import (
	"io"
	"log"
	"os"
	"path/filepath"
)

// SourceSkillsDir is the source skills directory (relative to runner binary)
// This is used to copy default skills to the user's .xiaoqinglong/skills on first init
var SourceSkillsDir = ""

// SourceConfigDir is the source config directory (relative to runner binary)
// This is used to copy default config files to the user's .xiaoqinglong/config on first init
var SourceConfigDir = ""

// BaseDirEnv is the environment variable for the base directory
const BaseDirEnv = "XQL_BASE_DIR"

// RunnerHomeEnv is the environment variable for RUNNER_HOME
// This allows complete profile isolation between different runner instances
const RunnerHomeEnv = "RUNNER_HOME"

// DefaultBaseDir is the default base directory name (in home directory)
const DefaultBaseDir = ".xiaoqinglong"

// GetBaseDir returns the unified base directory path
// Priority: RUNNER_HOME > XQL_BASE_DIR > ~/.xiaoqinglong
func GetBaseDir() string {
	// 1. Check RUNNER_HOME environment variable (highest priority, for profile isolation)
	if runnerHome := os.Getenv(RunnerHomeEnv); runnerHome != "" {
		if filepath.IsAbs(runnerHome) {
			return runnerHome
		}
		// Resolve relative paths relative to current working directory
		cwd, _ := os.Getwd()
		return filepath.Join(cwd, runnerHome)
	}

	// 2. Check XQL_BASE_DIR environment variable
	if baseDir := os.Getenv(BaseDirEnv); baseDir != "" {
		if filepath.IsAbs(baseDir) {
			return baseDir
		}
		// Resolve relative paths relative to home directory
		home, _ := os.UserHomeDir()
		return filepath.Join(home, baseDir)
	}

	// 3. Default to ~/.xiaoqinglong
	home, err := os.UserHomeDir()
	if err != nil {
		// Fallback to /tmp if home cannot be determined
		return filepath.Join("/tmp", DefaultBaseDir)
	}
	return filepath.Join(home, DefaultBaseDir)
}

// GetSkillsDir returns the skills directory path
func GetSkillsDir() string {
	return filepath.Join(GetBaseDir(), "skills")
}

// GetUploadsDir returns the uploads directory path
func GetUploadsDir() string {
	return filepath.Join(GetBaseDir(), "data", "uploads")
}

// GetReportsDir returns the reports directory path
func GetReportsDir() string {
	return filepath.Join(GetBaseDir(), "data", "reports")
}

// GetLogsDir returns the logs directory path
func GetLogsDir() string {
	return filepath.Join(GetBaseDir(), "logs")
}

// GetCheckpointsDir returns the checkpoints directory path
func GetCheckpointsDir() string {
	return filepath.Join(GetBaseDir(), "checkpoints")
}

// GetConfigDir returns the config directory path
func GetConfigDir() string {
	return filepath.Join(GetBaseDir(), "config")
}

// GetMemoryDir returns the memory directory path
func GetMemoryDir() string {
	return filepath.Join(GetBaseDir(), "memory")
}

// GetSessionMemoryDir returns the session memory directory path
func GetSessionMemoryDir(sessionID string) string {
	return filepath.Join(GetMemoryDir(), "sessions", sessionID)
}

// GetUserMemoryDir returns the user memory directory path
func GetUserMemoryDir(userID string) string {
	return filepath.Join(GetMemoryDir(), "users", userID)
}

// GetAgentMemoryDir returns the agent memory directory path
func GetAgentMemoryDir(agentID string) string {
	return filepath.Join(GetMemoryDir(), "agents", agentID)
}

// EnsureBaseDir ensures the base directory and all subdirectories exist
func EnsureBaseDir() error {
	dirs := []string{
		GetBaseDir(),
		GetSkillsDir(),
		GetUploadsDir(),
		GetReportsDir(),
		GetLogsDir(),
		GetCheckpointsDir(),
		GetConfigDir(),
		GetMemoryDir(),
	}

	for _, dir := range dirs {
		if err := os.MkdirAll(dir, 0755); err != nil {
			return err
		}
	}
	return nil
}

// Init ensures the base directory structure exists
// Should be called at runner startup
func Init() {
	if err := EnsureBaseDir(); err != nil {
		log.Fatalf("[xqldir] Warning: failed to create base directories: %v", err)
	} else {
		log.Printf("[xqldir] Base directory initialized: %s", GetBaseDir())
	}

	// 确保 skills 目录是正确的（不是 symlink），并复制默认 skills
	ensureSkillsDir()

	// 确保配置文件在 ~/.xiaoqinglong/config/ 目录
	ensureConfigFiles()
}

// ensureSkillsDir 确保 skills 目录存在且是真实目录（非 symlink），并复制默认 skills
func ensureSkillsDir() {
	skillsDir := GetSkillsDir()

	// 检查是否是 symlink
	if info, err := os.Lstat(skillsDir); err == nil && info.Mode()&os.ModeSymlink != 0 {
		log.Printf("[xqldir] Removing invalid symlink: %s -> %s", skillsDir, info.Name())
		if err := os.Remove(skillsDir); err != nil {
			log.Printf("[xqldir] Warning: failed to remove symlink: %v", err)
			return
		}
	}

	// 如果目录不存在，或者目录为空，复制默认 skills
	needsCopy := false
	if _, err := os.Stat(skillsDir); os.IsNotExist(err) {
		needsCopy = true
	} else {
		// 检查是否为空目录
		entries, err := os.ReadDir(skillsDir)
		if err != nil || len(entries) == 0 {
			needsCopy = true
		}
	}

	if needsCopy && SourceSkillsDir != "" {
		// 检查源目录是否存在
		if _, srcErr := os.Stat(SourceSkillsDir); srcErr == nil {
			log.Printf("[xqldir] Copying default skills from %s to %s", SourceSkillsDir, skillsDir)
			// 先删除可能存在的空目录
			os.RemoveAll(skillsDir)
			if err := copyDir(SourceSkillsDir, skillsDir); err != nil {
				log.Printf("[xqldir] Warning: failed to copy default skills: %v, creating empty dir", err)
				os.MkdirAll(skillsDir, 0755)
			}
		} else {
			log.Printf("[xqldir] Source skills dir not found: %s, creating empty skills dir", SourceSkillsDir)
			os.MkdirAll(skillsDir, 0755)
		}
	} else if needsCopy {
		os.MkdirAll(skillsDir, 0755)
	}
}

// ensureConfigFiles 确保配置文件在 ~/.xiaoqinglong/config/ 目录
func ensureConfigFiles() {
	configDir := GetConfigDir()

	// 创建配置目录（如果不存在）
	if err := os.MkdirAll(configDir, 0755); err != nil {
		log.Printf("[xqldir] Warning: failed to create config dir: %v", err)
		return
	}

	// 复制 skills-config.yaml（如果不存在）
	skillsConfigPath := filepath.Join(configDir, "skills-config.yaml")
	if _, err := os.Stat(skillsConfigPath); os.IsNotExist(err) {
		copySkillsConfig(configDir)
	}

	// 复制 config.yaml（如果不存在）
	configYamlPath := filepath.Join(configDir, "config.yaml")
	if _, err := os.Stat(configYamlPath); os.IsNotExist(err) {
		copyAgentConfig(configDir)
	}
}

func copySkillsConfig(configDir string) {
	// 从 SourceSkillsDir 复制 skills-config.yaml
	if SourceSkillsDir != "" {
		srcSkillsConfig := filepath.Join(SourceSkillsDir, "..", "skills-config.yaml")
		if _, err := os.Stat(srcSkillsConfig); err == nil {
			dstSkillsConfig := filepath.Join(configDir, "skills-config.yaml")
			if err := copyFile(srcSkillsConfig, dstSkillsConfig); err != nil {
				log.Printf("[xqldir] Warning: failed to copy skills-config.yaml: %v", err)
			} else {
				log.Printf("[xqldir] Copied skills-config.yaml to %s", dstSkillsConfig)
			}
		}
	}
}

// GetSourceConfigDir returns the source config directory
// This is used to find config files to copy to ~/.xiaoqinglong/config/
func GetSourceConfigDir() string {
	if SourceConfigDir != "" {
		return SourceConfigDir
	}

	// Fallback: 查找可执行文件同目录下的 config 目录
	execPath, err := os.Executable()
	if err != nil {
		return ""
	}
	execDir := filepath.Dir(execPath)
	localConfig := filepath.Join(execDir, "config")
	if _, err := os.Stat(localConfig); err == nil {
		return localConfig
	}

	// Fallback: 查找 dev 路径 (runner 在 build/bin/ 时，config 在 build/bin/config/)
	devConfig := filepath.Join(execDir, "..", "config")
	if _, err := os.Stat(devConfig); err == nil {
		return devConfig
	}

	return ""
}

func copyAgentConfig(configDir string) {
	srcDir := GetSourceConfigDir()
	if srcDir == "" {
		log.Printf("[xqldir] Source config dir not found, skipping config.yaml copy")
		return
	}

	srcConfig := filepath.Join(srcDir, "config.yaml")
	if _, err := os.Stat(srcConfig); err == nil {
		dstConfig := filepath.Join(configDir, "config.yaml")
		if err := copyFile(srcConfig, dstConfig); err != nil {
			log.Printf("[xqldir] Warning: failed to copy config.yaml: %v", err)
		} else {
			log.Printf("[xqldir] Copied config.yaml to %s", dstConfig)
		}
	}
}

// copyDir 复制目录（递归）
func copyDir(src, dst string) error {
	srcInfo, err := os.Stat(src)
	if err != nil {
		return err
	}

	if err := os.MkdirAll(dst, srcInfo.Mode()); err != nil {
		return err
	}

	entries, err := os.ReadDir(src)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())

		if entry.IsDir() {
			if err := copyDir(srcPath, dstPath); err != nil {
				return err
			}
		} else {
			if err := copyFile(srcPath, dstPath); err != nil {
				return err
			}
		}
	}
	return nil
}

// copyFile 复制文件
func copyFile(src, dst string) error {
	srcFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer srcFile.Close()

	dstFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer dstFile.Close()

	_, err = io.Copy(dstFile, srcFile)
	if err != nil {
		return err
	}

	// 复制权限
	srcInfo, _ := os.Stat(src)
	if srcInfo != nil {
		os.Chmod(dst, srcInfo.Mode())
	}

	return nil
}

File: backend/runner-langchain/app/utils/dir.py
"""Directory management following Go runner's xqldir design.

Environment variables (priority high to low):
1. RUNNER_HOME - Complete profile isolation
2. XQL_BASE_DIR - Base directory override
3. ~/.xiaoqinglong - Default

Directory structure:
~/.xiaoqinglong/
├── skills/           # User skills
├── config/           # Configuration files
├── logs/             # Log files
├── checkpoints/      # Execution checkpoints
├── memory/           # Memory storage
│   ├── sessions/    # Session memory
│   ├── users/       # User memory
│   └── agents/      # Agent memory
├── data/
│   ├── uploads/      # Uploaded files
│   └── reports/      # Generated reports
"""

import os
import shutil
import logging
from pathlib import Path

logger = logging.getLogger(__name__)

# Environment variable names (matching Go runner)
BASE_DIR_ENV = "XQL_BASE_DIR"
RUNNER_HOME_ENV = "RUNNER_HOME"

# Default base directory name
DEFAULT_BASE_DIR = ".xiaoqinglong"

# Source skills directory (relative to runner binary, for dev)
_SOURCE_SKILLS_DIR = ""


def get_base_dir() -> Path:
    """Get the unified base directory path.

    Priority:
    1. RUNNER_HOME (highest, for profile isolation)
    2. XQL_BASE_DIR
    3. ~/.xiaoqinglong (default)

    Returns:
        Path to base directory
    """
    # 1. Check RUNNER_HOME (highest priority)
    runner_home = os.environ.get(RUNNER_HOME_ENV)
    if runner_home:
        if os.path.isabs(runner_home):
            return Path(runner_home)
        # Resolve relative paths relative to cwd
        return Path.cwd() / runner_home

    # 2. Check XQL_BASE_DIR
    base_dir = os.environ.get(BASE_DIR_ENV)
    if base_dir:
        if os.path.isabs(base_dir):
            return Path(base_dir)
        # Resolve relative paths relative to home
        return Path.home() / base_dir

    # 3. Default to ~/.xiaoqinglong
    home = os.path.expanduser("~")
    if home == "~":
        # Fallback to /tmp if home cannot be determined
        return Path("/tmp") / DEFAULT_BASE_DIR
    return Path(home) / DEFAULT_BASE_DIR


def get_skills_dir() -> Path:
    """Get the skills directory path."""
    return get_base_dir() / "skills"


def get_uploads_dir() -> Path:
    """Get the uploads directory path."""
    return get_base_dir() / "data" / "uploads"


def get_reports_dir() -> Path:
    """Get the reports directory path."""
    return get_base_dir() / "data" / "reports"


def get_logs_dir() -> Path:
    """Get the logs directory path."""
    return get_base_dir() / "logs"


def get_checkpoints_dir() -> Path:
    """Get the checkpoints directory path."""
    return get_base_dir() / "checkpoints"


def get_config_dir() -> Path:
    """Get the config directory path."""
    return get_base_dir() / "config"


def get_memory_dir() -> Path:
    """Get the memory directory path."""
    return get_base_dir() / "memory"


def get_session_memory_dir(session_id: str) -> Path:
    """Get session memory directory."""
    return get_memory_dir() / "sessions" / session_id


def get_user_memory_dir(user_id: str) -> Path:
    """Get user memory directory."""
    return get_memory_dir() / "users" / user_id


def get_agent_memory_dir(agent_id: str) -> Path:
    """Get agent memory directory."""
    return get_memory_dir() / "agents" / agent_id


def ensure_base_dir() -> None:
    """Ensure base directory and all subdirectories exist."""
    dirs = [
        get_base_dir(),
        get_skills_dir(),
        get_uploads_dir(),
        get_reports_dir(),
        get_logs_dir(),
        get_checkpoints_dir(),
        get_config_dir(),
        get_memory_dir(),
    ]

    for d in dirs:
        try:
            d.mkdir(parents=True, exist_ok=True)
        except OSError as e:
            logger.warning("[xqldir] Failed to create directory %s: %s", d, e)


def _is_symlink(path: Path) -> bool:
    """Check if path is a symlink."""
    try:
        return path.is_symlink()
    except OSError:
        return False


def _copy_dir(src: Path, dst: Path) -> None:
    """Copy directory recursively."""
    if not src.exists():
        raise FileNotFoundError(f"Source directory not found: {src}")

    dst.mkdir(parents=True, exist_ok=True)

    for item in src.rglob("*"):
        if item.is_dir():
            dst_dir = dst / item.relative_to(src)
            dst_dir.mkdir(parents=True, exist_ok=True)
        else:
            dst_file = dst / item.relative_to(src)
            dst_file.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(item, dst_file)


def _copy_file(src: Path, dst: Path) -> None:
    """Copy a single file."""
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, dst)


def _is_dir_empty(path: Path) -> bool:
    """Check if directory is empty."""
    if not path.exists():
        return True
    try:
        return not any(path.iterdir())
    except OSError:
        return True


def ensure_skills_dir() -> None:
    """Ensure skills directory exists and copy default skills if needed.

    If skills dir is a symlink, remove it.
    If skills dir doesn't exist or is empty, copy from SourceSkillsDir.
    """
    skills_dir = get_skills_dir()

    # Check if it's a symlink
    if _is_symlink(skills_dir):
        logger.warning("[xqldir] Removing invalid symlink: %s", skills_dir)
        try:
            skills_dir.unlink()
        except OSError as e:
            logger.warning("[xqldir] Failed to remove symlink: %s", e)
            return

    # Check if needs copy
    needs_copy = False
    if not skills_dir.exists():
        needs_copy = True
    elif _is_dir_empty(skills_dir):
        needs_copy = True

    if needs_copy and _SOURCE_SKILLS_DIR:
        src = Path(_SOURCE_SKILLS_DIR)
        if src.exists():
            logger.info("[xqldir] Copying default skills from %s to %s", src, skills_dir)
            # Remove existing empty dir
            if skills_dir.exists():
                shutil.rmtree(skills_dir)
            try:
                _copy_dir(src, skills_dir)
            except Exception as e:
                logger.warning("[xqldir] Failed to copy default skills: %s", e)
                skills_dir.mkdir(parents=True, exist_ok=True)
        else:
            logger.info("[xqldir] Source skills dir not found: %s", src)
            skills_dir.mkdir(parents=True, exist_ok=True)
    elif needs_copy:
        skills_dir.mkdir(parents=True, exist_ok=True)


def ensure_config_files() -> None:
    """Ensure config files exist in config directory."""
    config_dir = get_config_dir()
    config_dir.mkdir(parents=True, exist_ok=True)

    # Copy skills-config.yaml if not exists
    skills_config = config_dir / "skills-config.yaml"
    if not skills_config.exists() and _SOURCE_SKILLS_DIR:
        src_config = Path(_SOURCE_SKILLS_DIR).parent / "skills-config.yaml"
        if src_config.exists():
            try:
                _copy_file(src_config, skills_config)
                logger.info("[xqldir] Copied skills-config.yaml to %s", skills_config)
            except Exception as e:
                logger.warning("[xqldir] Failed to copy skills-config.yaml: %s", e)


def init() -> None:
    """Initialize directory structure.

    Should be called at runner startup.
    """
    ensure_base_dir()
    logger.info("[xqldir] Base directory initialized: %s", get_base_dir())

    ensure_skills_dir()
    ensure_config_files()


def set_source_skills_dir(path: str) -> None:
    """Set the source skills directory (for development).

    This is used to copy default skills to the user's .xiaoqinglong/skills
    on first init.
    """
    global _SOURCE_SKILLS_DIR
    _SOURCE_SKILLS_DIR = path


def get_source_skills_dir() -> str:
    """Get the source skills directory."""
    return _SOURCE_SKILLS_DIR